File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/AddPlayerCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.domainModels.PlayerNotAssigned
import com.abaddon83.cardsGames.burracoGames.ports.{BurracoGameRepositoryPort, PlayerPort}
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.{Await, Future}

case class AddPlayerCmd(
                         gameIdentity: GameIdentity,
                         playerToAdd: PlayerNotAssigned
                       ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}


class AddPlayerHandler(
                        gameRepositoryPort: BurracoGameRepositoryPort
                      )
                      (implicit val ec: scala.concurrent.ExecutionContext) extends CommandHandler[AddPlayerCmd] {

  override def handleAsync(command: AddPlayerCmd): Future[Unit] = {
    assert(gameRepositoryPort.exists(command.gameIdentity), s"GameIdentity ${command.gameIdentity} doesn't exist")
    for {
      burracoGameWaitingPlayers <- gameRepositoryPort.findBurracoGameWaitingPlayersBy(command.gameIdentity)
    } yield gameRepositoryPort.save(burracoGameWaitingPlayers.addPlayer(command.playerToAdd))
  }

  override def handle(command: AddPlayerCmd): Unit = {
    import scala.concurrent.duration._
    import scala.language.postfixOps
    assert(gameRepositoryPort.exists(command.gameIdentity), s"GameIdentity ${command.gameIdentity} doesn't exist")
    val game = Await.result(gameRepositoryPort.findBurracoGameWaitingPlayersBy(command.gameIdentity), 5000 millis)
    gameRepositoryPort.save(game.addPlayer(command.playerToAdd))
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commands/AddPlayerCmd.kt
package com.abaddon83.burracoGame.commands

import com.abaddon83.burracoGame.domainModels.PlayerNotAssigned
import com.abaddon83.burracoGame.ports.BurracoGameRepositoryPort
import com.abaddon83.burracoGame.shared.games.GameIdentity
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*


data class AddPlayerCmd(
        val gameIdentity: GameIdentity,
        val playerToAdd: PlayerNotAssigned,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp): Command {

    constructor(gameIdentity: GameIdentity,
                playerToAdd: PlayerNotAssigned): this(
            gameIdentity = gameIdentity,
            playerToAdd = playerToAdd,
            commandId = UUID.randomUUID(),
            commandTimeStamp = Timestamp(System.currentTimeMillis())
    )
}


class AddPlayerHandler() : CommandHandler<AddPlayerCmd>, KoinComponent {

    private val repository: BurracoGameRepositoryPort by inject()


    override fun handle(command: AddPlayerCmd) {
        val burracoGameWaitingPlayers = runBlocking { repository.findBurracoGameWaitingPlayersBy(command.gameIdentity) }
        check(burracoGameWaitingPlayers != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        repository.save(burracoGameWaitingPlayers.addPlayer(command.playerToAdd))
    }

    override suspend fun handleAsync(command: AddPlayerCmd) {
        val burracoGameWaitingPlayers = repository.findBurracoGameWaitingPlayersBy(command.gameIdentity)
        check(burracoGameWaitingPlayers != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        repository.save(burracoGameWaitingPlayers.addPlayer(command.playerToAdd))
    }
}